.github/workflows/tests.yml

```diff
@@ -53,71 +56,224 @@ jobs:
 
       - name: Check upstream
         if: github.event_name == 'pull_request' && matrix.runtime-version == 'latest'
-        run: |
-          COMMIT="$(git log -n 1 --pretty=format:%s HEAD^2)"
-          if [[ "$COMMIT" == *"test-upstream"* ]]
-          then
-            export TEST_UPSTREAM="true"
-            # Put TEST_UPSTREAM into $GITHUB_ENV so it can be used in subsequent workflow steps
-            echo TEST_UPSTREAM=$TEST_UPSTREAM >> $GITHUB_ENV
-            echo "Using development version of dask and distributed"
-          fi
+        run: bash ci/scripts/check_upstream.sh
 
       - name: Install coiled-runtime
+        env:
+          COILED_RUNTIME_VERSION: ${{ matrix.runtime-version }}
+        run: bash ci/scripts/install_coiled_runtime.sh
+
+      - name: Export environment
+        env:
+          COILED_RUNTIME_VERSION: ${{ matrix.runtime-version }}
+          PYTHON_VERSION: ${{ matrix.python-version }}
+        run: bash ci/scripts/export_conda_env.sh
+
+      - name: Build Coiled Software Environment
```

Comment:
We're now building three times as many Coiled software environment. We're already building more than we need to since Coiled clusters always run on linux, there's no need to build a separate software environment for each os in the job matrix. This might be a good opportunity to move the Coiled software environment creation / deletion into their own jobs. Something like (psuedo-yaml below):

```yaml
jobs:
  software: <- build one software environment for each version of Python (with `runs-on: ["ubuntu-latest"]`)
  test_runtime: <- `needs: software`
  test_stability: <- `needs: software`
  test_benchmark: <- `needs: software`
  cleanup: <- We delete the Coiled software environments built in the `software` (`needs: [test_runtime, test_stability, test_benchmark]`)
```

That will both help us reduce the number of software environments we're building and cut down on repeated steps in each of the `test_*` jobs